### PR TITLE
Feature/notices

### LINKED ISSUE #311: Fatal Error if WooCommerce is not installed
#### Expected behavior and actual behavior.
If WooCommerce is missing we should print a notice warning the user, like this one:

> EBANX Payment Gateway for WooCommerce depends on the last version of WooCommerce to work
> [Link](https://github.com/ebanx/woocommerce-gateway-ebanx/blob/master/woocommerce-gateway-ebanx/includes/admin/views/html-notice-missing-woocommerce.php#L22)

But it is not working as it should. Instead it is throwing a Fatal Error.

#### Steps to reproduce the problem.
Install WC and EBANX Payment Gateway then deactive WC

#### Specifications like the version of the project, operating system, or hardware.
WC 2.6.14
EBANX Payment Gateway 1.2.1
Google Chrome 56.0.2924.87 (64-bit)
macOS Sierra

### PR DESCRIPTION
**Pull request by gpressutto5**
Fixes #311

### CODE REVIEW

**Comment by gpressutto5** [comment]
@SparK-Cruz What about this one?

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
+<?php
+
+if (!defined('ABSPATH')) {
+    exit;
+}
+
+class WC_EBANX_Notices_Notice {
```
Comment (cezarlz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Where is the comments? All methods are missing comments, we need to comment all of them. Please remember, this project is showing its face to the world 😆

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	private $allowed_types = array(
+			'error',
+			'warning',
+			'success',
+			'info'
+		);
+	private $is_dismissible = true;
+	private $view;
+
+	public function __construct() {
```
Comment (cezarlz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Missing comment here.

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+				$this->is_dismissible = $args[2];
+			case 2:
+				$this->with_type($args[1]);
+			case 1:
+				$this->message = $args[0];
+				break;
+		}
+	}
+
+	public function with_view($view) {
```
Comment (cezarlz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Missing comment here.

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		$this->view = $view;
+		return $this;
+	}
+
+	public function with_message($message) {
```
Comment (cezarlz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Missing comment here.

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		$this->message = $message;
+		return $this;
+	}
+
+	public function with_type($type) {
```
Comment (cezarlz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Missing comment here.

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+		if (!in_array($type, $this->allowed_types)) {
+			throw new InvalidArgumentException("Unknown notice type");
+		}
+		$this->type = $type;
+		return $this;
+	}
+
+	public function dismissible() {
```
Comment (cezarlz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Missing comment here.

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		$this->is_dismissible = true;
+		return $this;
+	}
+
+	public function persistent() {
```
Comment (cezarlz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Missing comment here.

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+		$this->is_dismissible = false;
+		return $this;
+	}
+
+	public function enqueue() {
```
Comment (cezarlz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Missing comment here.

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+	/**
+	 * Enqueues the notice to the WordPress hook
+	 * 
+	 * @return null
+	 * @throws Exception
+	 */
+	public function enqueue() {
+		if (isset($this->view)) {
+			include INCLUDES_DIR . 'admin/views/html-notice-'.$this->view.'.php';
```
Comment (SparK-Cruz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Wrap this in an action callback.

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+		}
+		if (is_null($this->message)) {
+			throw new Exception("You need to specify a message");
+		}
+		add_action('admin_notices', function () {
+			$classes = "notice notice-{$this->type}";
+			if ($this->is_dismissible) {
+				$classes .= ' is-dismissible';
+			}
+			$notice = "<div class='$classes'><p>{$this->message}</p></div>";
```
Comment (SparK-Cruz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Use local vars to be able to chain notices with a single instance.

**Comment by gpressutto5** [comment]
Roger that, Roger

File: woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php [hunk 11]
```diff
@@ -121,24 +121,32 @@ public function persistent() {
...
+			if ($is_dismissible) {
 				$classes .= ' is-dismissible';
 			}
-			$notice = "<div class='$classes'><p>{$this->message}</p></div>";
+			$notice = "<div class='$classes'><p>{$message}</p></div>";
 			echo $notice;
 		});
+		return $this;
 	}
 }
```
Comment (SparK-Cruz) [inline woocommerce-gateway-ebanx/includes/notices/class-wc-ebanx-notices-notice.php]: Add newline at the end of file

**Pull request merged**